explorer/explorer_js.go

```diff
@@ -3,48 +3,86 @@
 package explorer
 
 import (
+	"gioui.org/app"
 	"gioui.org/io/event"
 	"io"
 	"strings"
 	"syscall/js"
 )
 
-func listenEvents(_ event.Event) {}
+type explorer struct{}
 
-type File struct {
-	buffer                 js.Value
-	isWritable, isReadable bool
-	isClosed               bool
-	name                   string
-	index                  uint32
+func newExplorer(_ *app.Window) *explorer {
+	return &explorer{}
+}
+
+func (e *Explorer) listenEvents(_ event.Event) {
+	// NO-OP
+}
+
+func (e *Explorer) exportFile(name string) (io.WriteCloser, error) {
+	return newFileWriter(name), nil
+}
+
+func (e *Explorer) importFile(extensions ...string) (io.ReadCloser, error) {
+	// TODO: Replace with "File System Access API" when that becomes available on most browsers.
+	// BUG: Not work on iOS/Safari.
+
+	// It's not possible to know if the user closes the file-picker dialog, so an new channerl is needed.
+	r := make(chan result)
+
+	document := js.Global().Get("document")
+	input := document.Call("createElement", "input")
```

Comment:
Given that we document this:

> // It's a blocking call, you should call it on a separated goroutine. For most OSes, only one
// Import or Export, can happen at the same time, for each app.Window/Explorer.

Could we artificially enforce that constraint here? When you initiate a second file operation, assume that the previous one must be complete (or never will complete)? That would make it safe to reuse a single input instead of creating a new one each time.

Or does this block important use-cases? You're much more familiar with how these features are used than I am, so I'd love your thoughts on whether this is reasonable.

Reply:
The issue is if it's called while the file-dialog still opened. I'm not sure when it may happens, but JS is weird enough. 😅

Reply:
Okay, let's leave it as leaking for now. That really bugs me, but I don't know a good solution.